File: half-period-in-b-test.zig
// In a constant magnetic field the electron momentum change its direction
// to the opposite after a half period of the rotation. It is checked for
// different time steps here. Note that it is intended that the momentum pusher
// has single step error less than O((time step)^3).

const std = @import("std");
const core = @import("src/core.zig");
const Vec3 = core.Vec3;

// place here the momentum pusher you want to test
const pusher3P = @import("src/vay.zig").vay3P;

fn test_it(n_steps: i32) bool {
    const b = Vec3{ .x = 0.7, .y = 1.1, .z = 2.9 }; // just some magnetic field
    const p_initial = Vec3{ .x = -b.z, .y = 0, .z = b.x }; // the initial momentum; it is perpendicular to `b`

    const b_ampl = std.math.sqrt(core.norm2(Vec3, b));
    const gamma_initial = std.math.sqrt(1 + core.norm2(Vec3, p_initial));
    const period = 2 * std.math.pi * gamma_initial / b_ampl;
    const time_step = 0.5 * period / @intToFloat(f64, n_steps);
    const e = Vec3{ .x = 0, .y = 0, .z = 0 }; // electric field

    var i: i32 = 0;
    var p = p_initial;
    while (i < n_steps) : (i += 1) {
        p = pusher3P(p, -1, time_step, e, b);
    }

    const err = core.plus(Vec3, p_initial, p);
    const relative_error = std.math.sqrt(core.norm2(Vec3, err) / core.norm2(Vec3, p_initial));

    // true if the error is smaller than approximately 2 * (the estimate for error of midpoint methods)
    const is_ok = relative_error < std.math.pow(f64, time_step * b_ampl / gamma_initial, 2);
    return is_ok;
}

pub fn main() !void {
    const print = std.debug.print;

    const n_tests: i32 = 100;
    print("hi!\ndoing tests...\n", .{});

    var prng = std.rand.DefaultPrng.init(blk: {
        var seed: u64 = undefined;
        try std.os.getrandom(std.mem.asBytes(&seed));
        break :blk seed;
    });
    const rand = prng.random();

    var i: i32 = 0;
    while (i < n_tests) : (i += 1) {
        const n_steps = rand.intRangeAtMost(i32, 5, 200);
        if (!test_it(n_steps)) {
            print("test failed for n_steps = {d},\n", .{n_steps});
            break;
        }
    }
    if (i == n_tests) {
        print("successfully passed {d} tests,\n", .{n_tests});
    }
    print("bye!\n", .{});
}

File: src/bin/half-period-in-b-test.rs
// In a constant magnetic field the electron momentum change its direction
// to the opposite after a half period of the rotation. It is checked for
// different time steps here. Note that it is assumed here that the momentum
// pusher has single step error less than O((time step)^3).

use std::f64::consts::PI;
use rand::random_range;
use quoqqa_core::{Vec3, Norm, boris, vay3p, higuera_cary};
use quoqqa_core::pretty_print::{*};

fn test(pusher: fn(Vec3, f64, f64, Vec3, Vec3) -> Vec3, n_steps: i32) -> (bool, f64) {
    let b = Vec3{ x: 0.7, y: 1.1, z: 2.9 }; // just some magnetic field
    let p_initial = Vec3{ x: -b.z, y: 0.0, z: b.x }; // the initial momentum perpendicular to `b`

    let b_ampl = b.norm2().sqrt();
    let gamma_initial = (1.0 + p_initial.norm2()).sqrt();
    let period = 2.0 * PI * gamma_initial / b_ampl;
    let time_step = 0.5 * period / (n_steps as f64);
    let e = Vec3{ x: 0.0, y: 0.0, z: 0.0 }; // zero electric field

    let mut p = p_initial;
    for _i in 0..n_steps {
        p = pusher(p, -1.0, time_step, e, b);
    }

    let err = p_initial + p;
    let relative_error = (err.norm2() / p_initial.norm2()).sqrt();

    // true if the error is smaller than approximately 2 * (the estimate for error of midpoint methods)
    let estimate = (time_step * b_ampl / gamma_initial).powf(2.0);
    let is_ok = relative_error < estimate;
    (is_ok, relative_error / estimate)
}

fn main() {
    let n_tryes: i32 = 100;

    println!("Hi! Doing tests for pushers:");
    let mut ok: bool;
    let mut score: f64;

    for (pusher_name, function_name, pusher) in
    [("Boris", "boris", boris as fn(Vec3, f64, f64, Vec3, Vec3) -> Vec3),
    ("Vay", "vay3p", vay3p as fn(Vec3, f64, f64, Vec3, Vec3) -> Vec3),
    ("Higuera-Cary", "higuera_cary", higuera_cary as fn(Vec3, f64, f64, Vec3, Vec3) -> Vec3)] {
        print!("Testing {} pusher {}...", blue_bg(pusher_name), italic(function_name));
        ok = true;
        score = 0.0;
        for _i in 0..n_tryes {
            let n_timesteps = random_range(2..200);
            let (is_ok, error) = test(pusher, n_timesteps);
            score += error;
            if !is_ok {
                println!(" test {} for n_steps = {}", red_bg("failed!"), n_timesteps);
                ok = false;
                break;
            }
        }
        if ok {
            println!(" {} score={:.5}", green_bg("passed!"), score / (n_tryes as f64));
        }
    }

    println!("All done. Bye!");
}
